### PR TITLE
feat: keep track of fee sender addresses

### PR DESCRIPTION
**Pull request by mariocao**
## Motivation

This PR introduces several improvements to the fee handling system and request management in the core contracts:

1. Enable more flexible fee management by allowing different fee addresses for request, result, and batch fees
2. Allow duplicate unresolved post requests to update fees accordingly
3. Increase test coverage and improve code organization

These changes help prevent front-running attacks since posting requests or increasing fees will always use the highest fee amount.

## Explanation of Changes

The changes include:

1. **Fee Management Improvements**:
   - Allow different fee addresses for request, result, and batch fees
   - Use absolute values instead of increments for fees (clearer and safer)
   - Better fee distribution for multiple recipients

2. **Request Handling**:
   - Allow duplicate post requests if not yet resolved
   - Better request lifecycle management
   - Better validation of request states

3. **Storage Access**:
   - Changed storage functions from `private` to `internal` in base contracts

4. **Code Quality**:
   - Better comments and docs
   - Better function organization
   - Better NatSpec coverage

## Testing

All changes are tested:

1. **Fee Management Tests**:
   ```bash
   npx hardhat test test/core/SedaCoreV1.test.ts
   ```
   - Different fee address scenarios
   - Absolute fee handling
   - Fee distribution

2. **Request Handling Tests**:
   ```bash
   npx hardhat test
   ```
   - Duplicate request scenarios
   - Request lifecycle
   - Edge cases

3. **Coverage**:
   ```bash
   npx hardhat coverage
   ```
   - Better core coverage
   - Edge case testing
   - Better assertions

## Related PRs and Issues

N/A

### CODE REVIEW

File: contracts/interfaces/ISedaCore.sol [hunk 1]
```diff
@@ -40,17 +40,11 @@ interface ISedaCore is IResultHandler, IRequestHandler {
...
-    event FeesIncreased(
-        bytes32 indexed requestId,
-        uint256 additionalRequestFee,
-        uint256 additionalResultFee,
-        uint256 additionalBatchFee
-    );
+    /// @notice Emitted when fees are increased for an existing request
+    /// @param requestId The unique identifier of the request being updated
+    /// @param amount The absolute fee amount being increased
+    /// @param feeType The type of fee being increased (REQUEST, RESULT, or BATCH)
```
Comment (Thomasvdam) [inline contracts/interfaces/ISedaCore.sol:46]: Maybe these should also be `updated` instead of `increased`?

File: contracts/interfaces/ISedaCore.sol [hunk 2]
```diff
@@ -62,6 +56,14 @@ interface ISedaCore is IResultHandler, IRequestHandler {
     /// @notice Error thrown when attempting to set the timeout period to zero
     error InvalidTimeoutPeriod();
 
+    /// @notice Error thrown when attempting to increase fees but no fees were actually increased
```
Comment (Thomasvdam) [inline contracts/interfaces/ISedaCore.sol:59]: Same here

File: test/core/RequestHandler.test.ts [hunk 3]
```diff
@@ -72,11 +88,26 @@ describe('RequestHandler', () => {
...
+      await handler.postRequest(requests[0]);
+
+      // Second post of the same request should succeed and return the same ID
+      const duplicateRequestId = await handler.postRequest.staticCall(requests[0]);
+      expect(duplicateRequestId).to.equal(requestId);
+
+      // Verify the request data remains unchanged
+      const postedRequest = await handler.getRequest(requestId);
+      compareRequests(postedRequest, requests[0]);
+    });
```
Comment (Thomasvdam) [inline test/core/RequestHandler.test.ts:105]: This is the same as 46-59 right?

File: test/core/RequestHandler.test.ts [hunk 4]
```diff
@@ -43,27 +43,43 @@ describe('RequestHandler', () => {
...
 
-      await expect(handler.postRequest(requests[0]))
-        .to.be.revertedWithCustomError(handler, 'RequestAlreadyExists')
-        .withArgs(requestId);
+      // When posting a duplicate request with fees, it should succeed
+      const fee = ethers.parseEther('1');
+      await expect(
+        handler.postRequest.staticCall(requests[0], fee, 0, 0, { value: fee }),
+      ).to.be.revertedWithCustomError(handler, 'FeeManagerRequired');
     });
```
Comment (Thomasvdam) [inline test/core/RequestHandler.test.ts:71]: This test doesn't make sense? The description and comment say something different from the code.

**Review by Thomasvdam — APPROVED** [review]

**Comment by mariocao** [comment]
All comments addressed ^^

File: test/core/SedaCoreV1.test.ts [hunk 5]
```diff
@@ -102,36 +102,77 @@ describe('SedaCoreV1', () => {
...
 
-    return { prover, core, data };
+    return { prover, core, feeManager, data };
   }
 
-  describe('getPendingRequests', () => {
-    it('returns correct requests with pagination', async () => {
+  describe('postRequest', () => {
+    it('allows duplicate request submissions but does not repost', async () => {
       const { core, data } = await loadFixture(deployCoreFixture);
```
Comment (gluax) [inline test/core/SedaCoreV1.test.ts:110]: This test doesn't align with core contract right? We allow people to do this on the core contract?

File: test/core/SedaCoreV1.test.ts [hunk 6]
```diff
@@ -102,36 +102,77 @@ describe('SedaCoreV1', () => {
...
+      const pr2 = core.postRequest(data.requests[0]);
+
+      expect(pr2).not.to.emit(core, 'RequestPosted');
+
+      requests = await core.getPendingRequests(0, 10);
+      expect(requests.length).to.equal(1);
     });
 
-    it('returns zero requests if offset exceeds length', async () => {
+    it('prevents reposting a request after its result is posted', async () => {
```
Comment (gluax) [inline test/core/SedaCoreV1.test.ts:128]: Same here, believe this is allowed on core contract.

**Review by gluax — APPROVED** [review]

File: test/core/SedaCoreV1.test.ts (hunk 5 above)
Comment (mariocao) [reply test/core/SedaCoreV1.test.ts:110]: Yes, but only if it is not resolved. The idea is to deter front-running the `postRequest` (with low fees) because the legit `postRequest` will not "repost" (most probably wrong wording) but simply increase fees.

File: test/core/SedaCoreV1.test.ts (hunk 6 above)
Comment (mariocao) [reply test/core/SedaCoreV1.test.ts:128]: We agreed that we wanted to avoid recurring data request in this implementation to avoid edge cases with the solver. It is true that the protocol supports it.
